### PR TITLE
Confidence fields for indexing

### PR DESCRIPTION
**Pull request by dannymandel**

### CODE REVIEW

File: integration_tests/test_solr_search_results.py [hunk 1]
```diff
@@ -169,7 +171,8 @@ def _transformed_json_to_test_tuples() -> list[tuple]:
             current_tuple = (
                 key,
                 {
-                    "hasMaterialCategory": value.get("material").get("values")
+                    "hasMaterialCategory": value.get("material").get("values"),
+                    "hasMaterialCategoryConfidence": value.get("material").get("confidence")
```
Comment (dannymandel) [inline integration_tests/test_solr_search_results.py:175]: We now check both the indexed confidence values and the indexed string values

File: isamples_metadata/SESARTransformer.py [hunk 2]
```diff
@@ -301,17 +305,39 @@ def has_context_categories(self) -> typing.List[str]:
             material_type, primary_location_type
         )
 
-    def has_material_categories(self) -> typing.List[str]:
-        material = self._material_type()
-        if not material:
+    def _compute_material_prediction_results(self) -> typing.Optional[typing.List[PredictionResult]]:
```
Comment (dannymandel) [inline isamples_metadata/SESARTransformer.py:308]: Need to refactor this into something that will work generally across multiple category/confidence pairs.  Will do that work once we have the next model combo to integrate.

File: isamples_metadata/Transformer.py [hunk 3]
```diff
@@ -24,6 +24,10 @@ class Transformer(ABC):
 
     N2T_ARK_NO_HTTPS_PREFIX = f"{N2T_NO_HTTPS_PREFIX}ark:/"
 
+    RULE_BASED_CONFIDENCE = 1.0
```
Comment (dannymandel) [inline isamples_metadata/Transformer.py:27]: New confidence values as discussed on Wednesday -- 0…1 is machine-based, 1.0 is rule-based, 2.0 is human-entered.

File: isamples_metadata/Transformer.py [hunk 4]
```diff
@@ -134,21 +144,42 @@ def sample_sampling_purpose(self) -> str:
...
+            return None
+        confidences = []
+        for _ in category_list:
+            confidences.append(Transformer.RULE_BASED_CONFIDENCE)
+        return confidences
+
     @abstractmethod
     def has_context_categories(self) -> typing.List[str]:
         """Map from the source record into an iSamples context category"""
         pass
```
Comment (dannymandel) [inline isamples_metadata/Transformer.py:160]: New default implementations for all confidences -- will set things to 1.0 unless overridden.

File: scripts/migrations/solr/add_confidence_values.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
+import logging
```
Comment (dannymandel) [inline scripts/migrations/solr/add_confidence_values.py:1]: Solr migration to populate existing records with confidence values.

File: solr_schema_init/create_isb_core_schema.py [hunk 6]
```diff
@@ -128,8 +128,11 @@ def createCollection():
 createCopyField("description", "description_text")
 createCopyField("description", "searchText")
 createField("hasContextCategory", "string", True, True, None, True)
```
Comment (dannymandel) [inline solr_schema_init/create_isb_core_schema.py:130]: New solr fields to store the confidences.

**Review by datadavev — APPROVED** [review]
This all looks good. In general I think it would be good to have less code dependency on model property values (like the vocab terms), but the current approach works fine for now.

**Comment by dannymandel** [comment]
Could you explain a bit more about what you mean by that?  I'm not sure I understand.

**Comment by datadavev** [comment]
So for example MaterialCategoryMetaMapper hard codes different types of mappers tied to specific values. Could instead load the mappers load from a config like a json file, basically shifting the property specific values to config rather than code. The config might have entries like:
```json
[
  {
    "type":"endsWith",
    "input":["Rock"],
    "result":"Rock"
  },
  {
    "type":"equals",
    "input":["Siderite>Mineral", "Macrobiology>Other", "Organic Material"],
    "result":"Organic Material"
  }
]
``` 

Then MaterialCategoryMetaMapper might have something like loadJson() that populates its list of mappers. 

It shifts a bunch of stuff to config rather than code, but still functions the same so really the current approach is fine.

**Comment by dannymandel** [comment]
Got it.  Yeah, that makes sense and would be nice.  It might lower the barrier to entry for other people to write transformers.

**Pull request merged**